This patch file is from Qualcomm Innovation Center, Inc. and is provided
under the following license:

Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
SPDX-License-Identifier: BSD-3-Clause-Clear

From a02f654e1730d2a52b71cba43162bace181e3f13 Mon Sep 17 00:00:00 2001
From: ayukuma <quic_ayukuma@quicinc.com>
Date: Mon, 22 Jul 2024 15:12:21 +0530
Subject: [PATCH] ifindex_netifd_patch_02

ifindex issue:
1. after wlan device down and up, the ifindex may change
   but netifd init the ifindex value only once.
   now, set ifindex again whenever get DEV_EVENT_ADD event
2. after netifd get RTM_NEWLINK,
   it query ifindex but got 0 due race-condition.
   now, renew ifindex when found the ifindex is 0

Signed-off-by: Ayush Kumar<quic_ayukuma@quicinc.com>
---
 bridge.c       | 16 +++++++++++++++
 device.c       | 39 ++++++++++++++++++++++++++++++++++---
 device.h       |  3 ++-
 system-linux.c | 53 +++++++++++++++++++++++++++++++++++++-------------
 4 files changed, 93 insertions(+), 18 deletions(-)

diff --git a/bridge.c b/bridge.c
index cc25c51..e08c248 100644
--- a/bridge.c
+++ b/bridge.c
@@ -769,6 +769,22 @@ bridge_member_cb(struct device_user *dep, enum device_event ev)
 	case DEV_EVENT_AUTH_UP:
 		if (!bst->dev.active)
 			break;
+		//the device ifindex maybe change when wlan down and then up
+		//netifd now set ifindex only once, here to re-fresh the ifindex, otherwise bridge addif will failure
+		device_set_ifindex(dev, system_if_resolve(dev));
+
+		//netifd may get RTM_NETLINK before ifindex is generated,
+		//if the ifindex not ready, here we wait a few time to refresh the ifindex
+		if(dev->ifindex == 0) {
+			int try_number=0;
+			retry:
+				D(SYSTEM, "bridge_member_cb: sleep 50ms to wait %s.ifindex ready\n", bm->dev.dev->ifname);
+				usleep(50*1000);
+				device_set_ifindex(dev, system_if_resolve(dev));
+				try_number++;
+				if(try_number <=2 && dev->ifindex == 0)
+					goto retry;
+		}
 
 		if (bridge_enable_member(bm))
 			break;
diff --git a/device.c b/device.c
index f020cc0..84fe1fb 100644
--- a/device.c
+++ b/device.c
@@ -153,10 +153,27 @@ void device_stp_init(void)
 
 static int set_device_state(struct device *dev, bool state)
 {
+	int ret = 0;
+	int retry = 0;
+	D(DEVICE, "set_device_state: [%s]%d\n", dev->ifname, state);
 	if (state) {
 		/* Get ifindex for all devices being enabled so a valid  */
 		/* ifindex is in place avoiding possible race conditions */
 		device_set_ifindex(dev, system_if_resolve(dev));
+
+		//netifd may get RTM_NETLINK before ifindex is generated,
+		//if the ifindex not ready, here we wait a few time to refresh the ifindex
+		if(dev->ifindex == 0) {
+			int try_number=0;
+			retry:
+				D(DEVICE, "set_device_state: sleep 50ms to wait %s.ifindex ready\n", dev->ifname);
+				usleep(50*1000);
+				device_set_ifindex(dev, system_if_resolve(dev));
+				try_number++;
+				if(try_number <=4 && dev->ifindex == 0)
+					goto retry;
+		}
+
 		if (!dev->ifindex)
 			return -1;
 
@@ -166,9 +183,17 @@ static int set_device_state(struct device *dev, bool state)
 		dev->orig_settings.flags &= dev->settings.flags;
 		system_if_apply_settings(dev, &dev->settings, dev->settings.flags);
 
-		system_if_up(dev);
+		ret = system_if_up(dev);
+		while(ret < 0 && retry <=4 )
+		{
+			D(DEVICE, "set_device_state: system_if_up failed with ret=%d,errno:%d, sleep 50ms to retry\n", ret, errno);
+			retry++;
+			usleep(50*1000);
+			ret = system_if_up(dev);
+		}
+		D(DEVICE, "set_device_state: system_if_up result: %d, errno:%d\n", ret, errno);
 
-		system_if_apply_settings_after_up(dev, &dev->settings);
+		//system_if_apply_settings_after_up(dev, &dev->settings);
 	} else {
 		system_if_down(dev);
 		system_if_apply_settings(dev, &dev->orig_settings, dev->orig_settings.flags);
@@ -973,7 +998,7 @@ device_set_auth_status(struct device *dev, bool value, struct blob_attr *vlans)
 	device_broadcast_event(dev, DEV_EVENT_LINK_UP);
 }
 
-void device_set_present(struct device *dev, bool state)
+void device_set_present_private(struct device *dev, bool state)
 {
 	if (dev->sys_present == state)
 		return;
@@ -1471,5 +1496,13 @@ void device_hotplug_event(const char *name, bool add)
 	if (!dev || dev->type != &simple_device_type)
 		return;
 
+	if(dev->config_iface != NULL && add)  {
+	  if(dev->config_iface->state != IFS_UP && dev->config_iface->available == true)
+	  {
+            netifd_log_message(L_WARNING, "race condition: device %s was available but receive new hotplug on message\n", name);
+	    device_set_present(dev,false);
+	  }
+	}
+
 	device_set_present(dev, add);
 }
diff --git a/device.h b/device.h
index b2ea0fa..dd9bdcb 100644
--- a/device.h
+++ b/device.h
@@ -367,7 +367,8 @@ void device_add_user(struct device_user *dep, struct device *dev);
 void device_remove_user(struct device_user *dep);
 void device_broadcast_event(struct device *dev, enum device_event ev);
 
-void device_set_present(struct device *dev, bool state);
+void device_set_present_private(struct device *dev, bool state);
+#define device_set_present(dev, state) { netifd_log_message(L_DEBUG, "call device_set_present from %s,line %d\n",  __func__, __LINE__); device_set_present_private(dev, state);}
 void device_set_link(struct device *dev, bool state);
 void device_set_ifindex(struct device *dev, int ifindex);
 int device_set_ifname(struct device *dev, const char *name);
diff --git a/system-linux.c b/system-linux.c
index b6140cb..6adcd8d 100644
--- a/system-linux.c
+++ b/system-linux.c
@@ -706,8 +706,10 @@ system_device_update_state(struct device *dev, unsigned int flags, unsigned int
 static int cb_rtnl_event(struct nl_msg *msg, void *arg)
 {
 	struct nlmsghdr *nh = nlmsg_hdr(msg);
-	struct ifinfomsg *ifi = NLMSG_DATA(nh);
+	//struct ifinfomsg *ifi = NLMSG_DATA(nh);
 	struct nlattr *nla[__IFLA_MAX];
+	int link_state = 0;
+	char buf[10];
 	struct device *dev;
 
 	if (nh->nlmsg_type != RTM_NEWLINK)
@@ -721,7 +723,15 @@ static int cb_rtnl_event(struct nl_msg *msg, void *arg)
 	if (!dev)
 		return 0;
 
-	system_device_update_state(dev, ifi->ifi_flags, ifi->ifi_index);
+	//system_device_update_state(dev, ifi->ifi_flags, ifi->ifi_index);
+	if (!system_get_dev_sysfs("carrier", dev->ifname, buf, sizeof(buf)))
+		link_state = strtoul(buf, NULL, 0);
+
+	netifd_log_message(L_DEBUG, "cb_rtnl_event: iface %s link_state=%d", dev->ifname, link_state);
+//	if (dev->type == &simple_device_type)
+//		device_set_present(dev, true);
+
+	device_set_link(dev, link_state ? true : false);
 	return 0;
 }
 
@@ -732,7 +742,12 @@ handle_hotplug_msg(char *data, int size)
 	char *cur, *end, *sep;
 	int skip;
 	bool add;
+	bool isRndis=false;
 
+	if(strstr(data, "rndis")) {
+		isRndis = true;
+		netifd_log_message(L_DEBUG, "receive hotplug msg: %s\n", data);
+	}
 	if (!strncmp(data, "add@", 4) || !strncmp(data, "move@", 5))
 		add = true;
 	else if (!strncmp(data, "remove@", 7))
@@ -745,6 +760,9 @@ handle_hotplug_msg(char *data, int size)
 
 	for (cur = data + skip; cur < end; cur += skip) {
 		skip = strlen(cur) + 1;
+		if(isRndis) {
+			netifd_log_message(L_DEBUG, " detail data:%s\n", cur);
+		}
 
 		sep = strchr(cur, '=');
 		if (!sep)
@@ -767,9 +785,12 @@ handle_hotplug_msg(char *data, int size)
 	if (!subsystem || !interface)
 		return;
 
-	if (interface_old)
+	if (interface_old) {
+		netifd_log_message(L_DEBUG, "interface_old:%s\n", interface_old);
 		device_hotplug_event(interface_old, false);
+	}
 
+	netifd_log_message(L_DEBUG, "interface %s action %d\n", interface, add);
 	device_hotplug_event(interface, add);
 }
 
@@ -936,19 +957,23 @@ int system_bridge_addif(struct device *bridge, struct device *dev)
 	int tries = 0;
 	int ret;
 
-
-	for (tries = 0; tries < 3; tries++) {
-		ret = 0;
-		oldbr = system_get_bridge(dev->ifname, dev_buf, sizeof(dev_buf));
-		if (oldbr && !strcmp(oldbr, bridge->ifname))
-			break;
+retry:
+	ret = 0;
+	oldbr = system_get_bridge(dev->ifname, dev_buf, sizeof(dev_buf));
+	if (!oldbr || strcmp(oldbr, bridge->ifname) != 0) {
+
+		//retry set-ifindex to resolve racecondition that device not get ifindex
+		if(dev->ifindex == 0) {
+			device_set_ifindex(dev, system_if_resolve(dev));
+			D(SYSTEM, "system_bridge_addif: try get ifindex(new:%d)", dev->ifindex);
+		}
 
 		ret = system_bridge_if(bridge->ifname, dev, SIOCBRADDIF, NULL);
-		if (!ret)
-			break;
-
-		D(SYSTEM, "Failed to add device '%s' to bridge '%s' (tries=%d): %s\n",
-		  dev->ifname, bridge->ifname, tries, strerror(errno));
+		tries++;
+		D(SYSTEM, "Failed to add device '%s'/(ifindex=%d) to bridge '%s' (tries=%d): %s\n",
+		  dev->ifname, dev->ifindex, bridge->ifname, tries, strerror(errno));
+		if (tries <= 3)
+			goto retry;
 	}
 
 	if (dev->wireless)
-- 
2.17.1

